controllers/ndbserver_controller_helpers.go

```diff
@@ -0,0 +1,84 @@
+package controllers
+
+import (
+	"context"
+
+	ndbv1alpha1 "github.com/nutanix-cloud-native/ndb-operator/api/v1alpha1"
+	"github.com/nutanix-cloud-native/ndb-operator/common"
+	"github.com/nutanix-cloud-native/ndb-operator/common/util"
+	"github.com/nutanix-cloud-native/ndb-operator/ndb_api"
+	"github.com/nutanix-cloud-native/ndb-operator/ndb_client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Fetches all databases from the NDB API and converts them to
+// NDBServerDatabaseInfo type object (to be consumed by the NDBServer CR)
+func getNDBServerDatabasesInfo(ctx context.Context, ndbClient *ndb_client.NDBClient) (databases []ndbv1alpha1.NDBServerDatabaseInfo, err error) {
+	log := log.FromContext(ctx)
+	log.Info("Fetching and converting databases from NDB")
+
+	dbs, err := ndb_api.GetAllDatabases(ctx, ndbClient)
+	if err != nil {
+		log.Error(err, "NDB API error while fetching databases")
+		return
+	}
+	databases = make([]ndbv1alpha1.NDBServerDatabaseInfo, len(dbs))
+	for i, db := range dbs {
+		databaseInfo := ndbv1alpha1.NDBServerDatabaseInfo{
+			Name:   db.Name,
+			Id:     db.Id,
+			Status: db.Status,
+		}
+		if len(db.DatabaseNodes) > 0 {
+			databaseInfo.DBServerId = db.DatabaseNodes[0].DatabaseServerId
+			if len(db.DatabaseNodes[0].DbServer.IPAddresses) > 0 {
+				databaseInfo.IPAddress = db.DatabaseNodes[0].DbServer.IPAddresses[0]
+			}
+		}
+		databases[i] = databaseInfo
+	}
+	log.Info("Returning from ndbserver_controller_helpers.getNDBServerDatabasesInfo")
+	return
+}
+
+// Returns the NDBServerStatus after performing the following steps:
+// 1. Checks and fetch data if dbcounter is zero (we fetch data only when counter hits 0).
+// 2. TODO: Filter and set the required list of databases (we only want to store the databases managed by the operator).
+// 3. Update the counter value.
+func getNDBServerStatus(ctx context.Context, status *ndbv1alpha1.NDBServerStatus, ndbClient *ndb_client.NDBClient) *ndbv1alpha1.NDBServerStatus {
+	log := log.FromContext(ctx)
+	log.Info("Entered ndbserver_controller_helpers.getNDBServerStatus")
+
+	dbCounter := status.ReconcileCounter.Database
+	// 1. Fetch dbs only if dbcounter is 0
+	if dbCounter == 0 {
```

Comment:
Trying to understand why we fetch data only when the counter reaches to 0?

Reply:
The initial idea was to have the reconcile run every 'x' seconds (just to check if NDB connectivity is working fine etc) and update the databases, time machines and snapshots every 'dx', 'tx', 'sx' seconds. Since we're only dealing with databases initially, we simulate that behavior by maintaining this counter 'dbCounter'. Every time it reaches the `NDB_RECONCILE_DATABASE_COUNTER` (defined in constants, value: 4), it fetches the data for databases.